openpgp/internal/ecc/curve_info.go

```diff
@@ -1,97 +1,73 @@
 	"github.com/ProtonMail/go-crypto/brainpool"
+	"github.com/ProtonMail/go-crypto/openpgp/internal/encoding"
 )
 
 type SignatureAlgorithm uint8
 
-const (
-	ECDSA SignatureAlgorithm = 1
-	EdDSA SignatureAlgorithm = 2
-)
-
 type CurveInfo struct {
 	Name string
 	Oid *encoding.OID
-	Curve elliptic.Curve
-	SigAlgorithm SignatureAlgorithm
-	CurveType CurveType
+	Curve Curve
 }
 
-var curves = []CurveInfo{
+var Curves = []CurveInfo{
 	{
 		Name: "NIST curve P-256",
 		Oid: encoding.NewOID([]byte{0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07}),
-		Curve: elliptic.P256(),
-		CurveType: NISTCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(elliptic.P256(), NISTCurve),
 	},
 	{
 		Name: "NIST curve P-384",
 		Oid: encoding.NewOID([]byte{0x2B, 0x81, 0x04, 0x00, 0x22}),
-		Curve: elliptic.P384(),
-		CurveType: NISTCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(elliptic.P384(), NISTCurve),
 	},
 	{
 		Name: "NIST curve P-521",
 		Oid: encoding.NewOID([]byte{0x2B, 0x81, 0x04, 0x00, 0x23}),
-		Curve: elliptic.P521(),
-		CurveType: NISTCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(elliptic.P521(), NISTCurve),
 	},
 	{
 		Name: "SecP256k1",
 		Oid: encoding.NewOID([]byte{0x2B, 0x81, 0x04, 0x00, 0x0A}),
-		Curve: bitcurves.S256(),
-		CurveType: BitCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(bitcurves.S256(), BitCurve),
 	},
 	{
 		Name: "Curve25519",
 		Oid: encoding.NewOID([]byte{0x2B, 0x06, 0x01, 0x04, 0x01, 0x97, 0x55, 0x01, 0x05, 0x01}),
-		Curve: elliptic.P256(),// filler
-		CurveType: Curve25519,
-		SigAlgorithm: ECDSA,
+		Curve: NewCurve25519(),
 	},
 	{
 		Name: "Ed25519",
 		Oid: encoding.NewOID([]byte{0x2B, 0x06, 0x01, 0x04, 0x01, 0xDA, 0x47, 0x0F, 0x01}),
-		Curve: elliptic.P256(), // filler
-		CurveType: NISTCurve,
-		SigAlgorithm: EdDSA,
+		Curve: NewEd25519(),
 	},
 	{
 		Name: "Brainpool P256r1",
 		Oid: encoding.NewOID([]byte{0x2B, 0x24, 0x03, 0x03, 0x02, 0x08, 0x01, 0x01, 0x07}),
-		Curve: brainpool.P256r1(),
-		CurveType: BrainpoolCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(brainpool.P256r1(), BrainpoolCurve),
 	},
 	{
 		Name: "BrainpoolP384r1",
 		Oid: encoding.NewOID([]byte{0x2B, 0x24, 0x03, 0x03, 0x02, 0x08, 0x01, 0x01, 0x0B}),
-		Curve: brainpool.P384r1(),
-		CurveType: BrainpoolCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(brainpool.P384r1(), BrainpoolCurve),
 	},
 	{
 		Name: "BrainpoolP512r1",
 		Oid: encoding.NewOID([]byte{0x2B, 0x24, 0x03, 0x03, 0x02, 0x08, 0x01, 0x01, 0x0D}),
-		Curve: brainpool.P512r1(),
-		CurveType: BrainpoolCurve,
-		SigAlgorithm: ECDSA,
+		Curve: NewGenericCurve(brainpool.P512r1(), BrainpoolCurve),
 	},
 }
 
-func FindByCurve(curve elliptic.Curve) *CurveInfo {
-	for _, curveInfo := range curves {
-		if curveInfo.Curve == curve {
+func FindByCurve(curve Curve) *CurveInfo {
+	for _, curveInfo := range Curves {
+		if curveInfo.Curve.GetCurveType() == curve.GetCurveType() && curveInfo.Curve.GetCurveName() == curve.GetCurveName() {
```

Comment:
Apparently golang does a shallow compare, and therefore when using the generic curve two instances of the same curve are not equal (the type is the same but the pointer changes). Instead of comparing internals, that is difficult in an interface I decided to use unique naming.

Reply:
OK :+1: